### PR TITLE
Charlesmchen/redesign optimizations

### PR DESCRIPTION
**Pull request by charlesmchen-signal**
PTAL @michaelkirk 

Contains 3 separate changes in 3 commits.  They can be reviewed separately.

### CODE REVIEW

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageTextView.m [hunk 1]
```diff
@@ -7,6 +7,14 @@
 
 NS_ASSUME_NONNULL_BEGIN
 
+@interface OWSMessageTextView ()
+
+@property (nonatomic, nullable) NSValue *cachedSize;
```
Comment (charlesmchen-signal) [inline Signal/src/ViewControllers/ConversationView/Cells/OWSMessageTextView.m]: Cache the text view's size and ignore redundant updates.

File: Signal/src/util/DateUtil.m [hunk 2]
```diff
@@ -259,14 +259,20 @@ + (NSString *)formatTimestampShort:(uint64_t)timestamp
 
 + (NSString *)formatDateShort:(NSDate *)date
 {
+    OWSAssertIsOnMainThread();
     OWSAssert(date);
 
+    NSDate *now = [NSDate date];
+    NSInteger dayDifference = [self daysFromFirstDate:date toSecondDate:now];
+    BOOL dateIsOlderThanToday = dayDifference > 0;
+    BOOL dateIsOlderThanOneWeek = dayDifference > 6;
```
Comment (charlesmchen-signal) [inline Signal/src/util/DateUtil.m]: Date math is expensive; this cuts the cost in half.

**Comment by charlesmchen-signal** [comment]
> Contains 3 separate changes in 3 commits. They can be reviewed separately.

Make that 2 changes.

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageTextView.m [hunk 3]
```diff
@@ -61,6 +69,61 @@ - (BOOL)pointInside:(CGPoint)point withEvent:(UIEvent *_Nullable)event
...
+    self.cachedSize = nil;
+}
+
+- (void)setTextColor:(nullable UIColor *)textColor
+{
+    if ([NSObject isNullableObject:textColor equalTo:self.textColor]) {
+        return;
+    }
+    [super setTextColor:textColor];
+    self.cachedSize = nil;
```
Comment (michaelkirk-signal) [inline Signal/src/ViewControllers/ConversationView/Cells/OWSMessageTextView.m]: Text color changes the size?

File: Signal/src/util/DateUtil.m (hunk 2 above)
Comment (michaelkirk-signal) [reply Signal/src/util/DateUtil.m]: Hmm... any insight into why it's expensive?

It looks like this change would approximately cut it in half?

Comment (charlesmchen-signal) [reply Signal/src/util/DateUtil.m]: It's the usage of `NSCalendar` to slice the date into components that is surprisingly expensive. Due to the relative (to now) nature of these methods, we can't cache.

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageTextView.m (hunk 3 above)
Comment (charlesmchen-signal) [reply Signal/src/ViewControllers/ConversationView/Cells/OWSMessageTextView.m]: Yeah, we can skip that here.  Done, and added a comment.  Note that we still want this method; setting text color is not free.

**Pull request merged**